### PR TITLE
added in support for System.DateTime type for the DateTimeTransformer

### PR DESCRIPTION
**Pull request by michaelgsharp**
Added in support to the `DateTimeTransformer` to use columns of `DateTime` type.

Removed the thread safe caching as its faster to just query the native code multiple times then to deal with the thread safe cache.

### CODE REVIEW

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 1]
```diff
@@ -455,11 +465,17 @@ internal abstract class TypedColumn : IDisposable
...
+            internal readonly int StructSize;
 
             internal TypedColumn(string source, string prefix)
             {
                 Source = source;
                 Prefix = prefix;
+                IntPtrSize = IntPtr.Size;
+
+                // The native struct is 25 bytes + 8 size_t.
+                StructSize = 25 + (IntPtrSize * 8);
```
Comment (harishsk) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Is it possible to avoid hardcoded numbers? Would Marshal.Sizeof on the native struct work?

File: test/Microsoft.ML.Tests/Transformers/DateTimeTransformerTests.cs [hunk 2]
```diff
@@ -133,7 +133,50 @@ public void CanUseDateFromColumn()
             Done();
         }
 
-        [NotCentOS7FactAttribute]
+        [NotCentOS7Fact]
+        public void CanUseDateFromColumnDateTimeType()
+        {
+            // Future Date - 2025 June 30
```
Comment (harishsk) [inline test/Microsoft.ML.Tests/Transformers/DateTimeTransformerTests.cs]: If this code survives this date...will it still be the future? ;-)

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 3]
```diff
@@ -643,13 +655,15 @@ private sealed class Mapper : MapperBase
...
             private ConcurrentQueue<long> _oldestKeys;
+            private readonly DateTime _unixEpoc;
 
 #endregion
 
             public Mapper(DateTimeTransformer parent, DataViewSchema inputSchema) :
                 base(parent.Host.Register(nameof(Mapper)), inputSchema, parent)
             {
                 _parent = parent;
+                _unixEpoc = new DateTime(1970, 1, 1);
```
Comment (harishsk) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Nit: Typo for Epoch

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 4]
```diff
@@ -667,74 +681,68 @@ protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
...
+                var getter = input.GetGetter<TInput>(input.Schema[_parent._column.Source]);
+
+                ValueGetter<TTransformed> result = (ref TTransformed dst) =>
                 {
-                    long dateTime = default;
-                    getter(ref dateTime);
+                    TInput dateTimeColumn = default;
+                    getter(ref dateTimeColumn);
 
-                    if (!_cache.TryGetValue(dateTime, out TimePoint timePoint))
```
Comment (harishsk) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: I am curious about this deleted code. What was the cache for? Why is it no longer needed?

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 5]
```diff
@@ -148,7 +148,7 @@ internal DateTimeEstimator(IHostEnvironment env, Options options)
 
         public DateTimeTransformer Fit(IDataView input)
         {
-            return new DateTimeTransformer(_host, _options.Source, _options.Prefix, _options.Country);
+            return new DateTimeTransformer(_host, _options.Source, _options.Prefix, _options.Country, input);
         }
```
Comment (harishsk) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Nit: Since you are using only Schema in your class, would it be better to pass in the schema instead of input?

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 6]
```diff
@@ -745,7 +753,13 @@ protected override Delegate MakeGetter(DataViewRow input, int iinfo, Func<int, b
                 disposer = null;
 
                 // Have to add 1 to iinfo since the enum starts at 1
-                return Utils.MarshalInvoke(MakeGetter<int>, ((DateTimeEstimator.ColumnsProduced)iinfo + 1).GetRawColumnType(), input, iinfo);
+                // If input type is long, Posix time
+                if (input.Schema[_parent._column.Source].Type.RawType == typeof(long))
+                    return Utils.MarshalInvoke(MakeGetter<int, int>, new Type[] { typeof(long), ((DateTimeEstimator.ColumnsProduced)iinfo + 1).GetRawColumnType() }, input, iinfo);
+                // If input types is System.DateTime
+                else
```
Comment (harishsk) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Would an assert be better than a comment?

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 7]
```diff
@@ -643,13 +655,15 @@ private sealed class Mapper : MapperBase
             private readonly DateTimeTransformer _parent;
             private ConcurrentDictionary<long, TimePoint> _cache;
             private ConcurrentQueue<long> _oldestKeys;
+            private readonly DateTime _unixEpoc;
```
Comment (eerhardt) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: This should be a `static readonly` field. No need to new one up every time.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs [hunk 8]
```diff
@@ -667,74 +681,68 @@ protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
...
-                                long key;
-                                while (!_oldestKeys.TryDequeue(out key)) { }
-                                while (!_cache.TryRemove(key, out TimePoint removedValue)) { }
-                            }
-                        }
-                    }
+                    long dateTime;
+
+                    // Type already in Posix format
+                    if(typeof(TInput) == typeof(long))
```
Comment (eerhardt) [inline src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: This type checking should happen outside of the getter delegate. There is no need to pay this price every time this method is called. See other transformer implementations for examples.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 1 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: If I remember correctly I tried that and had issues with it, but let me try again.

Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Oh, the issue was that the native struct returns pointers to a string buffer and my struct uses strings directly. I can use Marshal.SizeOf, but then I need to define a second struct that exactly matches the native code. Would you prefer that approach?

File: test/Microsoft.ML.Tests/Transformers/DateTimeTransformerTests.cs (hunk 2 above)
Comment (michaelgsharp) [reply test/Microsoft.ML.Tests/Transformers/DateTimeTransformerTests.cs]: Good point. I'll remove the "Future" comment.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 4 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: I was trying to optimize by caching the retrieved value. Since this code can run in parallel though, the cache was thread safe. After doing some testing during this change it turned out in this case the thread safe cache was slower then just calling the native code extra times.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 5 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Good idea. Changed it to take just the schema.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 6 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: I was just trying to document the code execution path. If the type is not a supported type then an error will be thrown before it gets to this point.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 7 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Changed it as suggested.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 3 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Good catch. Fixed.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 8 above)
Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Refactored my MakeGetter and moved the type checking out of the delegate.

File: src/Microsoft.ML.Featurizers/DateTimeTransformer.cs (hunk 1 above)
Comment (harishsk) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Yes, I would prefer that approach. The presence of a struct that matches the native code makes the contract explicit and if something were to change on the native side, the corresponding change would be easier to identify.

---
In reply to: [368129495](https://github.com/dotnet/machinelearning/pull/4661#discussion_r368129495) [](ancestors = 368129495)

Comment (michaelgsharp) [reply src/Microsoft.ML.Featurizers/DateTimeTransformer.cs]: Sounds good. I have added in another struct that exactly matches the native code and am using that to get the size of the struct.
